### ISSUE TITLE
[debops.gitlab] exception on first run when using an external (non-localhost) PostgreSQL server

### ISSUE THREAD

**Issue by myrdd** [debops/debops#437]
**tl;dr** see fix at the bottom

## steps to reproduce

* run `debops.postgresql_server` on `your-postgresql-server.localhost`
* set `postgresql__server` to `your-postgresql-server.localhost`
* comment out [this `no_log` line in `ansible/roles/debops.postgresql/tasks/main.yml`](https://github.com/debops/debops/blob/4ea3e81a71bd16418cdb4b5a9494d8eaefcdae97/ansible/roles/debops.postgresql/tasks/main.yml#L83)
* run the [example `gitlab.yml` playbook](https://github.com/debops/debops/blob/4ea3e81a71bd16418cdb4b5a9494d8eaefcdae97/ansible/playbooks/service/gitlab.yml)

you'll get this exception:

<pre>
TASK [debops.postgresql : Create PostgreSQL roles] ****************************************
fatal: [{server name}]: FAILED! => {"msg": "<b>'dict object' has no attribute u'postgresql'</b>"}
</pre>

I've tracked down the problem:

## the issue

The `Create PostgreSQL roles` task uses these variables:

https://github.com/debops/debops/blob/4ea3e81a71bd16418cdb4b5a9494d8eaefcdae97/ansible/roles/debops.postgresql/tasks/main.yml#L75-L78

`postgresql__dependent_roles` is defined in the example playbook:

https://github.com/debops/debops/blob/4ea3e81a71bd16418cdb4b5a9494d8eaefcdae97/ansible/playbooks/service/gitlab.yml#L70-L73

`gitlab__postgresql__dependent_roles` is defined like this:

https://github.com/debops/debops/blob/4ea3e81a71bd16418cdb4b5a9494d8eaefcdae97/ansible/roles/debops.gitlab/defaults/main.yml#L853-L861

`gitlab_database_password`:

https://github.com/debops/debops/blob/4ea3e81a71bd16418cdb4b5a9494d8eaefcdae97/ansible/roles/debops.gitlab/defaults/main.yml#L303-L304

and finally `gitlab_database_password_path`:

https://github.com/debops/debops/blob/4ea3e81a71bd16418cdb4b5a9494d8eaefcdae97/ansible/roles/debops.gitlab/defaults/main.yml#L293-L297

There you can see that `gitlab_database_password_path` uses

* `ansible_local[gitlab__database].delegate_to`
* `ansible_local[gitlab__database].port`
* `gitlab_database_user`

However, `ansible_local[gitlab__database]` is not defined yet, because 

https://github.com/debops/debops/blob/4ea3e81a71bd16418cdb4b5a9494d8eaefcdae97/ansible/roles/debops.postgresql/tasks/main.yml#L246

will be run **at the end** of the `debops.postgresql`, but is needed earlier.

## further details

If `postgresql__server == "localhost"` the exception won't occur, because the `debops.postgresql_server` creates the `postgresql.fact` _before_ the `gitlab.yml` playbook is even run.

## how to fix?

I don't know what's the preferred way to fix this issue. Currently I'm overriding `gitlab_database_*` variables in the playbook like this:

```
  vars:
    gitlab_database_server: '{{ ansible_local[gitlab__database].server
                                if (ansible_local|d() and ansible_local[gitlab__database]|d() and
                                    ansible_local[gitlab__database].server|d())
                                else postgresql__server }}'
    gitlab_database_port: '{{ ansible_local[gitlab__database].port
                              if (ansible_local|d() and ansible_local[gitlab__database]|d() and
                                  ansible_local[gitlab__database].port|d())
                              else postgresql__port }}'
    gitlab_database_password_path: '{{ secret + "/" + gitlab__database + "/" +
                                       gitlab_database_server + "/" + gitlab_database_port +
                                       "/credentials/" + gitlab_database_user +
                                       "/password" }}'
```

and it works.

**Comment by drybjed** [comment]
Thanks for the bug report. I think that the best way to fix this would be to move the Ansible local facts tasks earlier in the `debops.postgresql` task list. This way the facts should be available at the time the role tries to manage the databases, roles, etc.

Can you check if this issue is resolved when you move the last three tasks in `debops.postgresql` just before the "Create PostgreSQL roles" task?

**Comment by myrdd** [comment]
> Can you check if this issue is resolved when you move the last three tasks in debops.postgresql just before the "Create PostgreSQL roles" task?

sure, I'll do

**Comment by myrdd** [comment]
it does fix the issue

**Comment by drybjed** [comment]
Great, looking forward to a PR. I don't think this will break anything else.
